packages/kbn-analytics/src/reporter.ts

```diff
@@ -79,36 +80,45 @@ export class Reporter {
     }
   }
 
-  public reportUiStats(
+  public reportUiStats = (
     appName: string,
-    type: UiStatsMetric['type'],
+    type: UiStatsMetricType,
     eventNames: string | string[],
     count?: number
-  ) {
+  ) => {
     const metrics = wrapArray(eventNames).map(eventName => {
-      if (this) this.log(`${type} Metric -> (${appName}:${eventName}):`);
+      this.log(`${type} Metric -> (${appName}:${eventName}):`);
       const report = createUiStatsMetric({ type, appName, eventName, count });
       this.log(report);
       return report;
     });
     this.saveToReport(metrics);
-  }
+  };
+
+  public reportUserAgent = (appName: string) => {
+    this.log(`Reporting user-agent.`);
+    const report = trackUsageAgent(appName);
+    this.saveToReport([report]);
+  };
 
-  public async sendReports() {
+  public sendReports = async () => {
     if (!this.reportManager.isReportEmpty()) {
       try {
         await this.http(this.reportManager.report);
         this.flushReport();
+        this.retryCount = 0;
       } catch (err) {
+        this.retryCount = this.retryCount + 1;
         this.log(`Error Sending Metrics Report ${err}`);
+        if (this.retryCount > this.maxRetries) {
+          this.log(`Max retries reached; Flushing report.`);
+          this.flushReport();
+        } else if (this.reportManager.report.reportVersion !== ReportManager.REPORT_VERSION) {
+          this.log(`Report Version mismatch; Flushing report.`);
+          this.flushReport();
```

Comment:
I can't help but feel that the flow here could be cleaned up a little.
When the _catch_ contains more logic than the _try_ I get worried, as it makes it hard to tell _what is the dangerous code here?_.
If the _http_ is the dangerous part, then why do we have _flushReport_ in both rather than after? if _flushReport_ is dangerous, then why do we call it in the catch?

Can we make this a little clearer perhaps?

Reply:
Makes sense; I'll take another stab at it.